Colinanderson/play kml tour compose

colinanderson opened the pull request:
## Description
<!--
PR to add a new Kotlin sample _"SAMPLE_NAME"_ in `SAMPLE_CATEGORY` category.
-->
Change the play KML tour sample to a compose implementation

## Links and Data

issue: https://devtopia.esri.com/runtime/kotlin/issues/4995
- a [vTest](https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/) Job for this PR has been run
  - [x] link: https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/58/

## What To Review
- Review the code to make sure it is easy to follow like other samples on Android
- `README.md` and `README.metadata.json` files

## How to Test
Run the sample on the sample viewer or the repo.

<!-- OPTIONAL
## To Discuss
-->

<!-- OPTIONAL
## Screenshots
-->

TADraeseke commented on samples/play-kml-tour/src/main/java/com/esri/arcgismaps/sample/playkmltour/components/PlayKMLTourViewModel.kt:
```diff
@@ -0,0 +1,158 @@
+
+    // create a scene with the surface and KML layer
+    val arcGISScene =
+        ArcGISScene(BasemapStyle.ArcGISImagery).apply {
```
can probably bump this up a line

TADraeseke commented on samples/play-kml-tour/src/main/java/com/esri/arcgismaps/sample/playkmltour/components/PlayKMLTourViewModel.kt:
```diff
@@ -0,0 +1,158 @@
+    private var kmlTour: KmlTour? = null
+    private val kmlTourController = KmlTourController()
+
+    private val _statusFlow: MutableStateFlow<KmlTourStatus> = MutableStateFlow(KmlTourStatus.NotInitialized)
+    val statusFlow = _statusFlow.asStateFlow()
```
could these maybe more explictly named? This code gets read outside of an IDE on the website, so more verbose names can be helpful

_tourStatus or _tourStatusFlow maybe?

TADraeseke commented on samples/play-kml-tour/src/main/java/com/esri/arcgismaps/sample/playkmltour/screens/PlayKMLTourScreen.kt:
```diff
@@ -0,0 +1,136 @@
+
+                val padding = 8.dp
+
+                LinearProgressIndicator(progress = { progressFlow },
+                    modifier = Modifier.fillMaxWidth().padding(all = padding))
+                Text("${stringResource(R.string.tour_status)} ${tourStateToString(stateFlow)}",
+                    modifier = Modifier.padding(all = padding).align(Alignment.CenterHorizontally))
+                Row(modifier = Modifier.fillMaxWidth()) {
+                    Button(enabled = (stateFlow == KmlTourStatus.Paused || stateFlow == KmlTourStatus.Playing),
+                        modifier = Modifier.fillMaxWidth(0.5f).padding(all = padding), onClick = {
+                        viewModel.reset()
+                    }) {
```
We normally make the chain of modifer.padding.fill.etc.etc vertically so more like this:

```suggestion
                LinearProgressIndicator(
                progress = { progressFlow }, modifier = Modifier
                    .fillMaxWidth()
                    .padding(all = padding)
            )
            Text(
                "${stringResource(R.string.tour_status)} ${tourStateToString(stateFlow)}",
                modifier = Modifier
                    .padding(all = padding)
                    .align(Alignment.CenterHorizontally)
            )
            Row(
                modifier = Modifier.fillMaxWidth()
            ) {
                Button(enabled = (stateFlow == KmlTourStatus.Paused || stateFlow == KmlTourStatus.Playing),
                    modifier = Modifier
                        .fillMaxWidth(0.5f)
                        .padding(all = padding),
                    onClick = {
                        viewModel.reset()
                    }) {
```

TADraeseke reviewed (changes requested):
Looking good @colinanderson -- just some small things below

colinanderson commented:
@TADraeseke that should be all your comments addressed.

TADraeseke reviewed (approved):
LGTM!

colinanderson commented:
Updated the enable/disable of the play button to disable when the tour is not initialised. Also renamed classes from `KML` to` Kml` and some other renames based on https://github.com/Esri/arcgis-maps-sdk-kotlin-samples/pull/287#discussion_r1871864445.

New vTest - https://runtime-kotlin.esri.com/view/all/job/vtest/job/sampleviewer/62

01smito01 reviewed (approved):
Looks good, just a few suggestions.

colinanderson commented on samples/play-kml-tour/README.metadata.json:
```diff
@@ -31,8 +33,10 @@
     ],
     "title": "Play KML tour"
-}
+}
```
Added

colinanderson commented on samples/play-kml-tour/src/main/java/com/esri/arcgismaps/sample/playkmltour/components/PlayKmlTourViewModel.kt:
```diff
@@ -0,0 +1,155 @@
+                if (kmlTour != null) {
+                    collectKmlTourStatus(kmlTour!!)
+                    collectProgress(kmlTourController)
+                }
```
Changed this and another `if` to be more Kotlin style.

colinanderson commented on samples/play-kml-tour/src/main/java/com/esri/arcgismaps/sample/playkmltour/screens/PlayKmlTourScreen.kt:
```diff
@@ -0,0 +1,147 @@
+import androidx.compose.runtime.getValue
+import androidx.compose.ui.Alignment
+import androidx.compose.ui.res.painterResource
+import androidx.compose.ui.unit.dp
```
Done

colinanderson commented on samples/play-kml-tour/src/main/java/com/esri/arcgismaps/sample/playkmltour/screens/PlayKmlTourScreen.kt:
```diff
@@ -0,0 +1,147 @@
+    val kmlTourStatus by kmlTourStatusFlow.collectAsStateWithLifecycle(KmlTourStatus.NotInitialized)
+
+    val kmlTourProgressFlow = remember { viewModel.kmlTourProgressFlow }
+    val kmlTourProgress by kmlTourProgressFlow.collectAsStateWithLifecycle(0.0f)
```
Changed as discussed.

colinanderson commented:
@01smito01 I made some updates so please take another look - interested if the `let` usages are an improvement or not.

01smito01 reviewed (approved):
I'm ambivalent on whether the `let`s are improvements, but I like the `when`. Everything else looks fine.
